proposals/0TBD-improved-resource-typing.md

```diff
@@ -0,0 +1,529 @@
+
+ * Binding 0 is a descriptor table
+ * Binding 1 is 4 root constants
+ * Binding 2 is a root descriptor CBV
+
+In Vulkan this would map to push constants, where `buffer` maps to a
+`VkDeviceAddress` value from a buffer.
+
+
+#### Open Issue: Allow resource entries to be constructed in-place?
+
+The above example shows the specification of a descriptor table from an index
+in `SV_Constants`.
+The manual step of construction is somewhat awkward, and it might make sense
+to have a way to define those directly in storage.
+Just specifying a structure would not be enough, as it would be treated as a
+structure of multiple offsets, rather than a single offset to a contiguous
+set of resources.
+
+Something like the following might be desirable:
+
+```hlsl
+struct RootData {
+	DescriptorTable<DescriptorTableData> descriptorTable;
+    uint4 constants;
+    RawConstantBuffer<...> buffer;
+};
+```
+
+This would also allow applications to load/store descriptor table handles
+directly in the same way as individual resource types.
+
+
+#### Open Issue: Allow explicit offsets?
+
+When defining a descriptor table, the offsets can be manually specified for
+a subset of the entries, with other entries calculated manually, taking into
+account any manually specified offsets as relevant.
+It might be useful to have an attribute in resource entry structures
+indicating this same functionality for compatibility reasons if nothing else.
+
+For example:
+
+```hlsl
+struct DescriptorTableData {
+    Texture2D<float4> a;
+    Texture2D<float4> b;
+    [[resourceoffset(16)]]
+    ConstantBuffer<Something> c;
+    RWBuffer<float4> d;
+    ...
+};
+```
+
+In this example the offset for `c` would be equal to 16, and `d` would take
+an offset as the sum of 16 and the size of `c`.
+It may also be beneficial to have a rolling offset variant, where the value
+is added to the otherwise calculated offset.
+
+Potentially this could be applicable to POD in regular structs as well, to
+enable more control over struct padding.
+
+
+### Raw Buffer types
+
+In the above example, a root descriptor (in that case a CBV) is passed in.
+When using a root descriptor in DirectX, these are mapped to buffers when
+using registers.
+However, declaring a standard buffer type directly here would result in the
+assumption of a 32-bit index into the heap, not a 64-bit root descriptor VA.
+
+New resource types are provided that can be declared to access a root
+descriptor VA:
+
+ * `RawConstantBuffer<typename T>`
+ * `RawStructuredBuffer<typename T = float4, uint64_t Size = UINT64_MAX>`
+ * `RawRWStructuredBuffer<typename T = float4, uint64_t Size = UINT64_MAX>`
+ * `RawByteAddressBuffer<uint64_t Size = UINT64_MAX>`
+ * `RawRWByteAddressBuffer<uint64_t Size = UINT64_MAX>`
+ * `RawRasterizerOrderedBuffer<typename T = float4, uint64_t Size = UINT64_MAX>`
+ * `RawRasterizerOrderedByteAddressBuffer<uint64_t Size = UINT64_MAX>`
+ * `RawRaytracingAccelerationStructure`
```

Comment:
Capturing feedback from myself and others: The usage of "raw" is a quite overloaded term, with the D3D *API* taking a "raw buffer" to be the equivalent of an HLSL `ByteAddressBuffer`. We need a different term for that.

Reply:
Yea, this was always a placeholder anyway. There's also some overlap with what LLVM calls "raw" which might confuse anyone working on clang.

I am wildly open to suggestions here.

Reply:
The only idea that comes to my mind, which I don't love, is "VAConstantBuffer", since you know, it's come from a VA. Maybe we call that one the backup plan. It's probably better than "Raw" though.

Reply:
I've made the change to call them "VA\<type>" instead, we can still haggle on the name, but at least it's not "Raw" anymore.